lib/graphql/internal_representation/rewrite.rb

```diff
@@ -45,6 +45,12 @@ def validate(context)
         visitor[Nodes::Field].enter << -> (ast_node, prev_ast_node) {
           parent_node = @nodes.last
           node_name = ast_node.alias || ast_node.name
+          
+          return :_skip if ast_node.directives.any? do |directive|
+            directive.name == 'skip' && directive.arguments.any? {|arg| arg.value == true } ||
+            directive.name == 'include' && directive.arguments.any? {|arg| arg.value == false }
+          end
```

Comment:
It is separate. I included it in this PR because after the document validation change for `@skip` and `@include`, it exposes some cases where graphql-ruby will improperly merge these directives. That error was prevented before by the validation (unless executing with `validation: false`. Happy to split into a separate PR, because as you said, the particular implementation deserves careful thought. From our end, the main thing we're concerned with is correctness, with optimization and efficiency being secondary (but still important).